### DISCUSSION TITLE
Modern Language Features

### DISCUSSION THREAD

**Discussion by yhack** [duckdb/duckdb#15170] [Ideas]
I began experimenting with DuckDB for a new project that requires fairly complex business logic. I'm very impressed DuckDB's Client APIs, documentation, and query features, but I haven't seen support for certain basic capabilities that facilitate building complex products.

### Namespaces, Imports, and Multiple Files
General-purpose programming languages such as Python, Java, and Rust support multiple files organized in a directory tree. Identifiers can be imported between files and aliased. Circular imports are automatically detected and execution order is inferred. To achieve similar behavior with DuckDB I found I could technically use multiple files, but I had to manually execute them in the right order. I didn't see any way to import views from one file into another. Instead, all references except those in the default schema required a fully-qualified name, which resulted in very long identifiers.  I also found myself missing encapsulation features. Some views needed to be built in steps, so I created intermediate views. These views should be considered private to the namespace, not exposed to the whole application, so I found myself looking for ways to denote this, such as starting the name with an underscore. CTEs are helpful, but they don't work when the intermediate queries need to be used in more than one final query or when I need to unit test them.

### Unit Tests
Unit tests are critical to the development of robust applications. The development documentation says as much regarding DuckDB itself. Oddly, the documentation says nothing about unit testing DuckDB queries written for applications. Wouldn't it make sense to make unit testing a first-class citizen? A step towards this could be to publish sqllogictest with the client libraries. As another modern example to look at is the [Rego](https://www.openpolicyagent.org/docs/latest/policy-testing/) policy language. It is based on Datalog, a cousin of SQL, yet it has unit testing built in.

### Summary
In summary, I'm suggesting that DuckDB support basic features of other modern languages that facilitate building large and complex products that are well structured and bug free, specifically the following:

- built-in handling of multiple files
- imports
- import aliases
- preference for encapsulation over globals
- built-in unit test support
